Resolves #263. exponentiate_pauli_sum, and test.

cbosborn opened the pull request.

willzeng commented on pyquil/paulis.py:
```diff
@@ -682,6 +682,26 @@ def exp_wrap(param):
+    Returns a function that maps all substituent PauliTerms and sums them into a program. NOTE: Use
+    this function with care if substituent PauliTerms do not commute.
+
+    :param PauliSum pauli_sum: Pauli terms to exponentiate.
```
`Pauli terms` --> `PauliSum`

ntezak commented on pyquil/paulis.py:
```diff
@@ -682,6 +682,26 @@ def exp_wrap(param):
     return exp_wrap
 
 
+def exponentiate_pauli_sum(pauli_sum):
```
I would prefer calling this `exponentiate_commuting_pauli_sum` as it is otherwise not correct/misleading.

ntezak replied:
Alternatively we can call it `exponentiate_pauli_terms(pauli_terms)` and then have some math in the documentation explain what unitary this corresponds to, i.e.

    exponentiate_pauli_terms([p1, p2]) == exp(1j*p2) * exp(1j*p1)

or something like that.

ecp-rigetti replied:
I super-agree with this.

ampolloreno replied:
:+1:

cbosborn replied:
I agree as well. Will make these changes.

ntezak commented on pyquil/tests/test_paulis.py:
```diff
@@ -330,6 +331,13 @@ def test_exponentiate():
 
 
+def test_exponentiate_pauli_sum():
+    pauli_sum = PauliSum([PauliTerm('X', 0, 0.5), PauliTerm('Z', 0, 0.5)])
```
This unit test uses non-commuting Pauli-terms, so please modify according to your decision regarding the name of this function and how to document it.

cbosborn commented:
@ntezak I've implemented the solution that assumes commuting Pauli terms and modified the test. As discussed, we can consider a more general implementation in the future, with the right checks and documentation.

ntezak commented:
@cbosborn last request, can you rebase off master?

cbosborn commented:
👍

stevenheidel commented on pyquil/paulis.py:
```diff
@@ -696,6 +696,26 @@ def exp_wrap(param):
+    fns = [exponential_map(term) for term in pauli_sum]
+
+    def combined_exp_wrap(param):
+        return sum([f(param) for f in fns], Program())
```
IMO the syntax Program([f(param) for f in fns]) is clearer than adding the second argument to sum but it's not a big deal

cbosborn replied:
I agree, changed.

willzeng replied:
oh yeah that is better. Maybe we should use that in some of the other code snippets too.
